accounts-db/src/accounts_db.rs

```diff
@@ -3674,18 +3690,50 @@ impl AccountsDb {
             "if filtering for incremental snapshots, then snapshots should be enabled",
         );
 
-        purges_zero_lamports.retain(|pubkey, (slot_account_infos, _ref_count)| {
+        purges_zero_lamports.retain(|pubkey, (slot_account_infos, ref_count)| {
             // Only keep purges_zero_lamports where the entire history of the account in the root set
             // can be purged. All AppendVecs for those updates are dead.
-            for (slot, _account_info) in slot_account_infos.iter() {
-                if let Some(store_count) = store_counts.get(slot) {
-                    if store_count.0 != 0 {
-                        // one store this pubkey is in is not being removed, so this pubkey cannot be removed at all
+            let mut force_allow_purge = false;
+            if slot_account_infos.len() == 1
+                && slot_account_infos.first().unwrap().1.is_zero_lamport()
+                && ref_count == &1
+            {
+                // `ref_count` is modified based on assumptions of what will happen in this active clean. Clean is in the middle of figuring out what will be
+                // dropped. We cannot rely on `ref_count` here. We need to check the actual index. ref_count may really be 2 and we may be hoping to drop the older element from the slot list.
+                // but, if that drop gets cancelled, then we would now resurrect an older version of this account in the index.
+                self.accounts_index.scan(
+                    [pubkey].into_iter(),
```

Comment:
Can we collect all candidate pubkeys and do the index scan together?

Reply:
yes, but... first, we don't expect many of these to occur. So the overhead of N individual scan calls vs one call is negligible. Second, grouping them up would rework the ordering and behavior of this function, which includes early returns, for example. That adds a lot of complexity, imo.

Reply:
sounds good!